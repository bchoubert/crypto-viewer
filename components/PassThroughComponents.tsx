/* eslint-disable import/prefer-default-export */
import { Image as ImageNative } from 'react-native';

export const Image = ImageNative;
