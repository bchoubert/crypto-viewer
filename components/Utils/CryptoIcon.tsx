import React, { memo } from 'react';
import { Image, StyleSheet, View } from 'react-native';

import CryptoCurrencyIconsMap from './CryptoCurrencyIconsMap';


const styles = StyleSheet.create({
  iconContainer: {
    marginRight: 10,
    borderRadius: 10,
  },
  icon: {
    width: 40,
    height: 40,
    maxHeight: 40,
  },
});


const CryptoIcon = ({ code }) => {
  return (
    <View style={{ ...styles.iconContainer, backgroundColor: CryptoCurrencyIconsMap[code]?.color || '#333333' }}>
      <Image style={styles.icon} source={CryptoCurrencyIconsMap[code]?.icon}  />
    </View>
  );
}

export default memo(CryptoIcon);
