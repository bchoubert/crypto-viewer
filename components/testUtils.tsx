/* eslint-disable import/no-extraneous-dependencies */
import { render, RenderAPI } from '@testing-library/react-native';
import '@testing-library/jest-dom/extend-expect';
import React, { ReactElement } from 'react';

import NavigationProvider from '../contexts/NavigationProvider';
import { SettingsContext } from '../contexts/SettingsProvider';
import ThemeProvider from '../contexts/ThemeProvider';
import TranslationProvider from '../contexts/TranslationProvider';

import { defaultSettings } from '../models/SettingsType';

jest.useFakeTimers();

export const enrichProviders = (node: ReactElement): ReactElement => (
  <SettingsContext.Provider
    value={{
      settings: defaultSettings,
      changeSettings: jest.fn(),
    }}
  >
    <TranslationProvider>
      <ThemeProvider>
        <NavigationProvider>
          {node}
        </NavigationProvider>
      </ThemeProvider>
    </TranslationProvider>
  </SettingsContext.Provider>
);

// eslint-disable-next-line max-len
export const renderNode = (node: ReactElement): RenderAPI => render(enrichProviders(node));
