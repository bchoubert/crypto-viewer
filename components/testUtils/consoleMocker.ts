/* eslint-disable no-console */
const consoleMocker = {
  consoleError: null,

  errorsToSuppress: [
    'Failed prop type',
  ],

  mockConsole: () => {
    consoleMocker.consoleError = console.error;
    console.error = (message: string) => {
      let printMessage = true;

      consoleMocker.errorsToSuppress.forEach((errorToSuppress: string) => {
        if (message.includes(errorToSuppress)) {
          printMessage = false;
        }
      });

      if (printMessage) {
        consoleMocker.consoleError?.(message);
      }
    };
  },

  restoreConsole: () => {
    if (consoleMocker.consoleError) {
      console.error = consoleMocker.consoleError;
    }
  },
};

export default consoleMocker;
