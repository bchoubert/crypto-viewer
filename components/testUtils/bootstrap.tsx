/* eslint-disable no-console */
/* eslint-disable import/no-extraneous-dependencies */
import { render, RenderAPI } from '@testing-library/react-native';
import '@testing-library/jest-dom/extend-expect';
import React, { ReactElement } from 'react';
import * as PassThoughComponents from '../PassThroughComponents';

import NavigationProvider from '../../contexts/NavigationProvider';
import { SettingsContext } from '../../contexts/SettingsProvider';
import ThemeProvider from '../../contexts/ThemeProvider';
import TranslationProvider from '../../contexts/TranslationProvider';

import { defaultSettings } from '../../models/SettingsType';
import consoleMocker from './consoleMocker';

jest.useFakeTimers();

export const enrichProviders = (node: ReactElement): ReactElement => (
  <SettingsContext.Provider
    value={{
      settings: defaultSettings,
      changeSettings: jest.fn(),
    }}
  >
    <TranslationProvider>
      <ThemeProvider>
        <NavigationProvider>
          {node}
        </NavigationProvider>
      </ThemeProvider>
    </TranslationProvider>
  </SettingsContext.Provider>
);

// eslint-disable-next-line max-len
export const renderNode = (node: ReactElement): RenderAPI => render(enrichProviders(node));

beforeAll(() => {
  consoleMocker.mockConsole();
  // eslint-disable-next-line no-extend-native
  Date.prototype.getTimezoneOffset = jest.fn(() => 0);
});

afterAll(() => {
  consoleMocker.restoreConsole();
  jest.resetAllMocks();
});

jest.spyOn(PassThoughComponents, 'Image');
(PassThoughComponents.Image as unknown as jest.Mock).mockImplementation(() => (<div>IMAGE</div>));
