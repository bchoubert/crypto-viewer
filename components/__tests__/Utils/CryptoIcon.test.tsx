import React from 'react';
import { Image } from '../../PassThroughComponents';
import { renderNode } from '../../testUtils/bootstrap';
import CryptoIcon from '../../Utils/CryptoIcon';

describe('<CryptoIcon />', () => {
  it('Prints correctly', () => {
    const { toJSON } = renderNode(
      <CryptoIcon
        code="btc"
      />,
    );

    expect(toJSON()).toMatchSnapshot();

    expect(Image).toHaveBeenCalledTimes(1);
  });

  it('No error even if crypto code is not existing', () => {
    renderNode(
      <CryptoIcon
        code="non_existing_crypto"
      />,
    );

    expect(Image).toHaveBeenCalledTimes(2);
  });
});
