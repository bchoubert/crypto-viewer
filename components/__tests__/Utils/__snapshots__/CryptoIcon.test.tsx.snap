// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CryptoIcon /> Prints correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F7931B",
      "borderRadius": 10,
      "marginRight": 10,
    }
  }
>
  <div>
    IMAGE
  </div>
</View>
`;
